File: src/main/java/no/nav/melosys/eessi/models/sed/MedlemskapTypeResolver.java
package no.nav.melosys.eessi.models.sed;

import java.util.Arrays;
import java.util.EnumMap;
import java.util.List;
import java.util.stream.Collectors;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import com.fasterxml.jackson.databind.DatabindContext;
import com.fasterxml.jackson.databind.JavaType;
import com.fasterxml.jackson.databind.jsontype.TypeIdResolver;
import lombok.extern.slf4j.Slf4j;
import no.nav.melosys.eessi.models.sed.medlemskap.Medlemskap;
import no.nav.melosys.eessi.models.sed.medlemskap.impl.*;

@Slf4j
class MedlemskapTypeResolver implements TypeIdResolver {

    private static final Class<? extends Medlemskap> DEFAULT_CLASS = NoType.class;
    private static final List<String> SED_TYPES_STRING = Arrays.stream(SedType.values()).map(SedType::name).collect(Collectors.toList());

    private static final EnumMap<SedType, Class<? extends Medlemskap>> mapping = new EnumMap<>(SedType.class);

    static {
        mapping.put(SedType.A001, MedlemskapA001.class);
        mapping.put(SedType.A002, MedlemskapA002.class);
        mapping.put(SedType.A003, MedlemskapA003.class);
        mapping.put(SedType.A004, MedlemskapA004.class);
        mapping.put(SedType.A005, MedlemskapA005.class);
        mapping.put(SedType.A006, MedlemskapA006.class);
        mapping.put(SedType.A007, MedlemskapA007.class);
        mapping.put(SedType.A008, MedlemskapA008.class);
        mapping.put(SedType.A009, MedlemskapA009.class);
        mapping.put(SedType.A010, MedlemskapA010.class);
        mapping.put(SedType.A011, MedlemskapA011.class);
        mapping.put(SedType.A012, MedlemskapA012.class);
    }

    private JavaType sedType;

    @Override
    public void init(JavaType javaType) {
        this.sedType = javaType;
    }

    @Override
    public String idFromValue(Object o) {
        return null;
    }

    @Override
    public String idFromValueAndType(Object o, Class<?> aClass) {
        return null;
    }

    @Override
    public String idFromBaseType() {
        return null;
    }

    @Override
    public JavaType typeFromId(DatabindContext databindContext, String s) {
        Class<?> type;
        if (SED_TYPES_STRING.contains(s)) {
            type = mapping.get(SedType.valueOf(s));
        } else {
            type = DEFAULT_CLASS;
        }

        return databindContext.constructSpecializedType(sedType, type);
    }

    @Override
    public String getDescForKnownTypeIds() {
        return null;
    }

    @Override
    public JsonTypeInfo.Id getMechanism() {
        return JsonTypeInfo.Id.NAME;
    }
}

File: melosys-eessi-app/src/main/java/no/nav/melosys/eessi/models/sed/MedlemskapTypeResolver.kt
package no.nav.melosys.eessi.models.sed

import com.fasterxml.jackson.annotation.JsonTypeInfo
import no.nav.melosys.eessi.models.SedType
import no.nav.melosys.eessi.models.sed.medlemskap.impl.*
import tools.jackson.databind.DatabindContext
import tools.jackson.databind.JavaType
import tools.jackson.databind.jsontype.TypeIdResolver

class MedlemskapTypeResolver : TypeIdResolver {
   private lateinit var sedType: JavaType

    override fun init(javaType: JavaType) {
        this.sedType = javaType
    }

    override fun idFromValue(ctxt: DatabindContext, value: Any): String? = null

    override fun idFromValueAndType(ctxt: DatabindContext, value: Any, suggestedType: Class<*>): String? = null

    override fun idFromBaseType(ctxt: DatabindContext): String? = null

    override fun typeFromId(context: DatabindContext, id: String): JavaType {
        val type = mapping[SedType.valueOf(id)] ?: DEFAULT_CLASS
        return context.constructSpecializedType(sedType, type)
    }

    override fun getDescForKnownTypeIds(): String? = null

    override fun getMechanism(): JsonTypeInfo.Id = JsonTypeInfo.Id.NAME

    companion object {
        private val DEFAULT_CLASS = NoType::class.java
        private val mapping = mapOf(
            SedType.A001 to MedlemskapA001::class.java,
            SedType.A002 to MedlemskapA002::class.java,
            SedType.A003 to MedlemskapA003::class.java,
            SedType.A004 to MedlemskapA004::class.java,
            SedType.A005 to MedlemskapA005::class.java,
            SedType.A006 to MedlemskapA006::class.java,
            SedType.A007 to MedlemskapA007::class.java,
            SedType.A008 to MedlemskapA008::class.java,
            SedType.A009 to MedlemskapA009::class.java,
            SedType.A010 to MedlemskapA010::class.java,
            SedType.A011 to MedlemskapA011::class.java,
            SedType.A012 to MedlemskapA012::class.java,
            SedType.H010 to MedlemskapH010::class.java
        )
    }
}
